pkg/ha/ha_service.go

```diff
@@ -5,245 +5,170 @@
 
+// Service contains the lease state for all prometheus clusters
+// and logic for determining if a specific sample should
+// be allowed to be inserted. Also it keeps the lease state
+// up to date by periodically refreshing it from the database.
 type Service struct {
 	state               *sync.Map
-	lockClient          haLockClient
-	leaseTimeout        time.Duration
-	leaseRefresh        time.Duration
+	leaseClient         client.LeaseClient
 	leaderChangeLocks   *sync.Map
 	syncTicker          util.Ticker
 	currentTimeProvider func() time.Time
 }
 
-func NewHAService(dbClient pgxconn.PgxConn) (*Service, error) {
+// NewHaService constructs a new HA service with the supplied
+// database connection and the default sync interval.
+// The sync interval determines how often the leases for
+// all clusters are refreshed from the database.
+func NewHAService(dbClient pgxconn.PgxConn) *Service {
 	return NewHAServiceWith(dbClient, util.NewTicker(haSyncerTimeInterval), time.Now)
 }
 
-func NewHAServiceWith(dbClient pgxconn.PgxConn, ticker util.Ticker, currentTimeFn func() time.Time) (*Service, error) {
-	lockClient := newHaLockClient(dbClient)
-	timeout, refresh, err := lockClient.readLeaseSettings(context.Background())
-	if err != nil {
-		return nil, fmt.Errorf("could not create HA state: %#v", err)
-	}
+// NewHaServiceWith constructs a new HA service with the supplied
+// database connection, ticker and a current time provide function.
+// The ticker determines when the lease states for all clusters
+// are refreshed from the database, the currentTimeFn determines the
+// current time, used for deterministic tests.
+func NewHAServiceWith(dbClient pgxconn.PgxConn, ticker util.Ticker, currentTimeFn func() time.Time) *Service {
+	lockClient := client.NewHaLockClient(dbClient)
 
 	service := &Service{
 		state:               &sync.Map{},
-		lockClient:          lockClient,
-		leaseTimeout:        timeout,
-		leaseRefresh:        refresh,
+		leaseClient:         lockClient,
 		leaderChangeLocks:   &sync.Map{},
 		syncTicker:          ticker,
 		currentTimeProvider: currentTimeFn,
 	}
 	go service.haStateSyncer()
-	return service, nil
+	return service
 }
 
-func (h *Service) haStateSyncer() {
+// haStateSyncer periodically synchronizes the in-memory
+// lease states with latest values from the database and initiates
+// a leader change if the conditions are met.
+func (s *Service) haStateSyncer() {
 	for {
-		h.syncTicker.Wait()
-		h.state.Range(func(c, s interface{}) bool {
+		s.syncTicker.Wait()
+		s.state.Range(func(c, l interface{}) bool {
 			cluster := fmt.Sprint(c)
-			ss := castToState(s)
-			stateView := ss.Clone()
-			latestState, err := h.syncLockStateFromDB(stateView.LeaseStart, stateView.MaxTimeSeenLeader, cluster, stateView.Leader)
+			lease := l.(*state.Lease)
+			stateBeforeUpdate := lease.Clone()
+			err := lease.UpdateFromDB(s.leaseClient, stateBeforeUpdate.Leader, stateBeforeUpdate.LeaseStart, stateBeforeUpdate.MaxTimeSeenLeader)
```

Comment:
Totally optional (and not sure it's a good idea): the fact that the function UpdateFromDB takes so much states that's part of lease itself might imply it's signature should be
`lease.UpdateFromDB(s.leaseClient)`  or rather `lease.UpdateLease(s.leaseClient)`

for CheckLease below we change to:
`lease.CheckLease(s.leaseClient))` where the code block starting with `if !maxT.Before(leaseView.LeaseUntil) {` below moves into lease.CheckLease and uses the times set by `UpdateMaxSeenTime`. That does imply lease should track minT for leader in addition to maxT (so minT should be passed into and set in `UpdateMaxSeenTime`). 

Again, maybe this is a bad idea.

Reply:
I think it's not ok, 
1. We need to track minT, and there may be edge cases for when it needs/doesn't need to be updated
2. Sending the leader as an argument provides a good check that the state was not modified from another routine between Clone() and UpdateFromDB